### ISSUE TITLE
SFL4J warnings

### ISSUE THREAD

**Issue by SlimeDog** [plan-player-analytics/Plan#2438]
### Describe the issue

Plan 5.5-b1740 yields the following WARNings on start-up. Is there something I need to configure?

### Exceptions & Other Logs

```
[11:33:44] [Server thread/WARN]: SLF4J: No SLF4J providers were found.
[11:33:44] [Server thread/WARN]: SLF4J: Defaulting to no-operation (NOP) logger implementation
[11:33:44] [Server thread/WARN]: SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
[11:33:44] [Server thread/WARN]: SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions prior to 1.8.
[11:33:44] [Server thread/WARN]: SLF4J: Ignoring binding found at [jar:file:/home/ram/MC/plugins/Plan-5.5-b1740.jar!/plan/org/slf4j/impl/StaticLoggerBinder.class]
[11:33:44] [Server thread/WARN]: SLF4J: See http://www.slf4j.org/codes.html#ignoredBindings for an explanation.
```

### Plugin versions
On both Spigot and Paper 1.19
Plan 5.5-b1740

**Comment by AuroraLS3** [comment]
Ah this is because of the Sponge API 8 update, as the slf4j relocation rules were changed a bit, it'll be fixed before full release 🙂 - can be safely ignored.

**Comment by AuroraLS3** [comment]
I'm having some difficulties solving this. I think the issue is that Jetty is trying to load slf4j through ServiceLoader, but it's using wrong classloader (Thread context classloader instead of the plugin classloader).

This is what I've tried so far
- Updated slf4j to 2.0.0-alpha7, removed the last 2 messages from the error (about ignored binding)
- Stopped relocating slf4j, the message still appears on Bungeecord.

If anybody knows how to sort this out that would be great

**Comment by Kopo942** [comment]
One of the following logging providers needs to be added on the classpath for the warnings to disappear: slf4j-nop.jar, slf4j-simple.jar, slf4j-log4j12.jar, or slf4j-jdk14.jar.

**Comment by AuroraLS3** [comment]
I thought slf4j-nop was already included 🤔

**Comment by AuroraLS3** [comment]
Confirmed that it is a classloader issue.
Thread context classloader doesn't find the serviceprovider, but plugin class loader does (On bungeecord)
```
// Thread context
10:10:07 [INFO] plan.org.slf4j.spi.SLF4JServiceProvider (looking for)
10:10:07 [INFO] [] (found)
// After swapping to plugin class loader
10:10:07 [INFO] plan.org.slf4j.spi.SLF4JServiceProvider (looking for)
10:10:07 [INFO] [plan.org.slf4j.nop.NOPServiceProvider] (found)
```

<details>
<summary>code</summary>

```java
        ServiceLoader<SLF4JServiceProvider> serviceLoader = ServiceLoader.load(SLF4JServiceProvider.class);
        List<SLF4JServiceProvider> providerList = new ArrayList<>();
        for (SLF4JServiceProvider provider : serviceLoader) {
            providerList.add(provider);
        }
        System.out.println(SLF4JServiceProvider.class.getName());
        System.out.println(providerList.stream().map(SLF4JServiceProvider::getClass).map(Class::getName).collect(Collectors.toList()));
        System.out.println("...");
        System.out.println(NOPServiceProvider.class.getName());
        ClassLoader pluginClassLoader = getClass().getClassLoader();
        ThreadContextClassLoaderSwap.performOperation(pluginClassLoader, () -> {
            ServiceLoader<SLF4JServiceProvider> serviceLoader2 = ServiceLoader.load(SLF4JServiceProvider.class);
            List<SLF4JServiceProvider> providerList2 = new ArrayList<>();
            for (SLF4JServiceProvider provider : serviceLoader2) {
                providerList2.add(provider);
            }
            System.out.println(SLF4JServiceProvider.class.getName());
            System.out.println(providerList2.stream().map(SLF4JServiceProvider::getClass).map(Class::getName).collect(Collectors.toList()));
            System.out.println("...");
            System.out.println(NOPServiceProvider.class.getName());
        });
        
    public static void performOperation(ClassLoader usingClassLoader, Runnable operation) {
        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
        try {
            // Jetty uses Thread context classloader, so we need to change to plugin classloader where the ALPNProcessor is.
            Thread.currentThread().setContextClassLoader(usingClassLoader);

            operation.run();
        } finally {
            Thread.currentThread().setContextClassLoader(contextClassLoader);
        }
    }
    
    // Taken from LoggerFactory.class (slf4j)
    private static List<SLF4JServiceProvider> findServiceProviders() {
        ServiceLoader<SLF4JServiceProvider> serviceLoader = ServiceLoader.load(SLF4JServiceProvider.class);
        List<SLF4JServiceProvider> providerList = new ArrayList<>();
        for (SLF4JServiceProvider provider : serviceLoader) {
            providerList.add(provider);
        }
        return providerList;
    }
```

</details>

**Comment by AuroraLS3** [comment]
Removing slf4j from Plan jar entirely stops the warning on bukkit, bungeecord and velocity, but Plan crashes on nukkit with NoClassDefFoundError (as nukkit doesn't include slf4j). Fabric and sponge are still untested.

I haven't found a way to get the class inside thread context classloader yet - it is the classloader of the platform that is loading the plugin and changing context classloader to be plugin classloader in a static block sounds very bad (since there is no block that runs when classes finish loading when it could be changed back).

**Comment by Vankka** [comment]
PlanBungee#onEnable
```java
            Thread currentThread = Thread.currentThread();
            ClassLoader contextClassLoader = currentThread.getContextClassLoader();
            try {
                currentThread.setContextClassLoader(getClass().getClassLoader());
                system = component.system();
            } finally {
                currentThread.setContextClassLoader(contextClassLoader);
            }
```
This would be better than changing the context class loader permanently, there's still a off chance that one of the proxy's classes could be loaded with the context loader, but that should be fine, probably.

**Comment by AuroraLS3** [comment]
Ahh that works! I totally overlooked that Dagger doesn't initialize the object graph *before* calling `Component#system`. Thanks again Vankka!

**Comment by SlimeDog** [comment]
Confirmed fixed on Spigot and Paper with Plan 5.5 build 1791. Thanks.
